### DISCUSSION TITLE
Feature request/discussion: history limit for backups by ScheduledBackup

### DISCUSSION THREAD

**Discussion by georgebarbarosie** [cloudnative-pg/cloudnative-pg#1284] [Ideas]
Currently, `Backup` manifests generated by `ScheduledBackup` will accumulate indefinitely, requiring either manual cleanup or some type of external automation. I see a couple of ways to automate the cleanup of backups.
- similar to the history limits of k8s CronJob (`.spec.successfulJobsHistoryLimit` and `.spec.failedJobsHistoryLimit`)
- automatically based on retentionPolicy

There are advantage to both. The problem I see with the 2nd option is that the manager would have to replicate barman's window of recoverability logic. A mitigation for this would be to use the output of `barman-cloud-backup-delete` as the event triggering the deletion of the associated Backup.

**Comment by georgebarbarosie** [comment]
A little update on the current behaviour of the operator (version 1.18). I've initially filed this issue after asking in Slack about this functionality but after some testing I now think the cluster manager does something pretty close to what I was describing as mitigation above. I am pretty certain that even a manually created backup manifest gets purged alongside its object store presence. This is undocumented behaviour, which might cause some surprise but it actually makes perfect sense. So this feature request is pointless now, or should be transformed into a request for feature documentation rather than implementation.

**Comment by MichaluxPL** [comment]
I can confirm that both scheduled backups and on-demand backups are purged (both manifests and stored backup files) according to retention policy set.
I'm not sure if on-demand backups should drop into the same box with scheduled backups in terms of retention. Sometimes we do an on-demand backup to have additional copy for some purposes (from a particular place in time) and want to keep it for some time. So maybe a good solution would be to provide with some additional option to allow the user to choose if on-demand backups should be a subject to backup retention policy or not ?

**Comment by georgebarbarosie** [comment]
I am curious about how backup manifests are purged. Is `barman-cloud-backup-delete` informing the manager about which `backupId`s have been removed? This is the best case scenario, because this means manual backups (all of them, or potentially based on a manifest flag, only those marked by the user as archival backups) can be tagged on creation with `barman-cloud-backup-keep`. If my understanding of that command's effect is correct, this would ensure the desired effect, and neither the base backup in the object store nor the manifest would get deleted. I will be performing some tests to see if this is correct and get back here with the results.

**Comment by georgebarbarosie** [comment]
I tested the theory by manually marking base backups as archival using `barman-cloud-backup-keep` and I am highly confident now that the manager evicts `Backup` manifests based on information obtained from the `barman-cloud-backup-delete` output. This is great news.

I see as a low hanging fruit the addition of a field in the `Backup` manifest to change the archival status of the barman base backup (available targets are `standalone` and `full`).

**Comment by sxd** [comment]
Hi,

This looks like an idea that can be implemented in Barman rather than in the operator, what do you think @georgebarbarosie @MichaluxPL  ?

I'm turning this into discussion!

Best Regards
